### PR TITLE
Add E2E flows and FW architecture for Keyword Detection (KD) use case

### PR DESCRIPTION
**Pull request by lbetlej**
The Keyword Detection algorithm requires dedicated cross-domain support  that spans between SW & FW. The PR covers arch documentation for intro, E2E flows, FW topology and FW components functional scope to support  integration of a custom keyword detection algorithm. The SW/FW infrastructure support includes buffering of a data used for keyword  detection in FW and a succeeding  command for an additional processing on host CPU.
Power flows & scheduling considerations will be described separately.

### CODE REVIEW

File: howtos/kd_integration/images/sys-dev-states.pu [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+Sys has "S0" as high
+Sys has "S0ix" as low
+Sys has "S3" as none
+
+robust "Device" as Dev
+Dev has "D0" as high
+Dev has "D0ix" as low
+Dev has "D3" as none
+
+robust "WOV" as Wov
```
Comment (lgirdwood) [inline howtos/kd_integration/images/sys-dev-states.pu]: Do we want to remove all WOV references and use KD instead ?

Comment (lbetlej) [reply howtos/kd_integration/images/sys-dev-states.pu]: fixed. I've updated PR to remove all references to files that are in a draft state.

File: developer_guides/apps/kd_integration/images/kd-e2e-sequence-diagram.pu [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+Host system may enter the low power state
+end ref
+
+loop keyword detection algorithm \nexecuted on DSP
+ kda <- gpdma 
+end
+
+hnote over kda : keyword is detected
+
+fw <-- kda : notification on keyword detection
```
Comment (mrajwa) [inline developer_guides/apps/kd_integration/images/kd-e2e-sequence-diagram.pu]: I think we agreed to call it *event* on FW side.

Comment (lbetlej) [reply developer_guides/apps/kd_integration/images/kd-e2e-sequence-diagram.pu]: good catch, I've updated the diagram to keep terminology consistent

**Comment by lgirdwood** [comment]
@intelkevinputnam

**Comment by intelkevinputnam** [comment]
@lgirdwood taking a look now

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+End-2-End flows
+***************
+
+.. uml:: images/kd-e2e-sequence-diagram.pu
+   :caption: E2E flow for SW/FW components
+
+The fundamental assumption for the flow is that keyphrase detection sequence is
+controlled by user space component (application) opening and closing speech
+audio stream. The audio topology setup needs to happen before speech stream is
+opened. There is optional sequence to customize the keyword detection algorithm
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 64: "... There is an optional sequence that ..."

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+End-2-End flows
+***************
+
+.. uml:: images/kd-e2e-sequence-diagram.pu
+   :caption: E2E flow for SW/FW components
+
+The fundamental assumption for the flow is that keyphrase detection sequence is
+controlled by user space component (application) opening and closing speech
+audio stream. The audio topology setup needs to happen before speech stream is
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 63: "... audio topology setup needs to happen before the speech ..."

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+opened. There is optional sequence to customize the keyword detection algorithm
+by behavior by sending run-time parameters. Stream open and preparation phase
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 65: "... The Stream open and ..."

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+covers among others sending HW parameters to DAI and passing configuration
+parameters from topology to FW components. The DAPM events handlers are used to
+control a Keypharse Detector node of FW topology graph by audio driver. Once
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 68: "... driver. Once the"

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+the host. Once the buffer is drained the speech capture pipeline starts to work
+as a passthrough capture until is closed by user space application.
+
+FW topology
+***********
+
+.. uml:: images/kd-component-diagram.pu
+   :caption: Basic diagram for FW components topology
+
+The diagram provides overview of FW and HW components that play role in
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 80: "... provides an overview ..." "... that play a role in"

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+keyphrase detection flows. The components are organized in pipelines:
+
+1. Speech capture pipeline.
+
+   a) DMIC DAI configures hw interface to capture data from microphones.
+
+   b) Keyphrase Buffer Managrer is responsible for management of a data captured
+      by microphones. This includes control of a internal buffer for incoming
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 88: "... control of an internal ..."

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+      data and routing of incoming of incoming audio samples. The audio buffer
+      with historic audio data is implemented as a cyclic buffer. While
+      listeining to a keyphrase the component stores incoming data in internal
+      buffer and copies it to sink that leads toward keyword detecttor
+      component. On successful detection of a keyphrase the buffer is drained
+      during a burst transmission to a host. Once buffer is drained it starts
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 94: "... Once the buffer is drained ..."

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+      to work as a passthrough component on a capture pipeline.
+
+   c) Host component configures transport (over DMA) to host system. The
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 97: "... The host component ..." "... to the host system ..."

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+      component is  responsible for a transmission from local memory
+      (FW accessible) to remote (host CPU accessible) memory.
+
+
+2. Keyphrase Detector pipeline.
+
+   a) Channel selector is responsible for providing a single channel on input to
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 104: "... The channel selector ..." "... input to the"

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+      keyphrase detection algorithm. Decision on selected channel is made by
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 105: "... The decision of which channel to select is made ..."

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+      platform integrator. The component can accept parameters from a topology
+      file.
+
+   b) Keyphrase detection algorithm accepts audio frames and returns information
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 109: "... The keyphrase ..."

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+      if keyphrase is detected. Note that FW infrastructure allows to send FW
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 110: "if a keyphrase ..."

File: developer_guides/apps/kd_integration/kd-integration.rst [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+      event to Keyphrase Buffer Manager component if keyphrase is detected.
+      The component sends also notification to audio driver and implements
```
Comment (intelkevinputnam) [inline developer_guides/apps/kd_integration/kd-integration.rst]: Line 112: "The component also sends a ..."

**Pull request merged**